### PR TITLE
PP-11027 Require and use gateway account ID to fetch webhooks data

### PR DESCRIPTION
**Pull request by sfount**
Depends on https://github.com/alphagov/pay-webhooks/pull/820.

Until the data model across Pay shifts to being service-centric, use the
gateway account ID to allow consumers to fetch webhooks data
(this will allow multiple test card payment accounts to be
associated with a single service without causing any issues in the
webhooks mechanism or reporting on webhooks data, for the time being).

### CODE REVIEW

File: src/main/java/uk/gov/pay/webhooks/webhook/resource/WebhookResource.java [hunk 1]
```diff
@@ -166,17 +172,19 @@ public List<WebhookResponse> getWebhooks(@Parameter(example = "true", descriptio
...
                                              @QueryParam("override_service_id_restriction") boolean overrideServiceIdRestriction) {
-        if (service_id != null && overrideServiceIdRestriction) {
-            throw new BadRequestException("service_id not permitted when using override_service_id_restriction");
+        if ((service_id != null || gatewayAccountId != null) && overrideServiceIdRestriction) {
+            throw new BadRequestException("[service_id, gateway_account_id] not permitted when using override_service_id_restriction");
         }
 
-        if (service_id == null && !overrideServiceIdRestriction) {
-            throw new BadRequestException("either service_id or override_service_id_restriction query parameter must be provided");
+        if ((service_id == null || gatewayAccountId == null) && !overrideServiceIdRestriction) {
```
Comment (stephencdaly) [inline src/main/java/uk/gov/pay/webhooks/webhook/resource/WebhookResource.java:183]: Are there some tests for the various logic for this on different endpoints that we can add to to ensure that gateway_account_id is required as expected?

File: src/test/java/uk/gov/pay/webhooks/webhook/resource/WebhookResourceTest.java [hunk 2]
```diff
@@ -283,7 +290,21 @@ public void getWebhooksRequestWithLiveParamOnlyShould400() throws JsonProcessing
         Map<String, String> responseBody = objectMapper.readValue(response.readEntity(String.class), new TypeReference<>() {
         });
         assertThat(response.getStatus(), is(400));
-        assertThat(responseBody.get("message"), is("either service_id or override_service_id_restriction query parameter must be provided"));
+        assertThat(responseBody.get("message"), is("[service_id, gateway_account_id] or override_service_id_restriction query parameter must be provided"));
+    }
+
+    @Test
+    public void getWebhooksRequestWithLiveAndServiceWithoutGatewayAccoutnIdShould400() throws JsonProcessingException {
```
Comment (stephencdaly) [inline src/test/java/uk/gov/pay/webhooks/webhook/resource/WebhookResourceTest.java:297]:
```suggestion
    public void getWebhooksRequestWithLiveAndServiceWithoutGatewayAccountIdShould400() throws JsonProcessingException {
```

could we add a similar tests for `/v1/webhook/{webhookExternalId}`?

File: src/test/java/uk/gov/pay/webhooks/webhook/resource/WebhookResourceTest.java [hunk 3]
```diff
@@ -152,11 +155,12 @@ public void getWebhookByIdWhenWebhookExistsAndServiceIdIncorrect404() {
 
     @Test
     public void getWebhookByIdWhenDoesNotExist404() {
-        when(webhookService.findByExternalIdAndServiceId(any(String.class), any(String.class))).thenReturn(Optional.empty());
+        when(webhookService.findByExternalIdAndGatewayAccountId(any(String.class), any(String.class))).thenReturn(Optional.empty());
 
         var response = resources
                 .target("/v1/webhook/%s".formatted("aint_no_webhook"))
                 .queryParam("service_id", "aint_no_service_id")
```
Comment (james-peacock-gds) [inline src/test/java/uk/gov/pay/webhooks/webhook/resource/WebhookResourceTest.java:162]: similarly here shouldn't the service id be valid for the purposes of this test?

**Review by james-peacock-gds — APPROVED** [review]
LGTM

**Pull request merged**